### ISSUE TITLE
Interfacing MCU SD Card directly for improved upload speed and reliabilty

### ISSUE THREAD

**Issue by Atreidae** [guysoft/OctoPi#38]
Hey guys, First time I've posted in github, but hey gotta start somewhere.

I've been running OctoPi on my Hadron w/Rumba board for a little while now and I've noticed that with larger files octopi struggles to feed the printer the first few curves of a print and fails due to a serial error

Uploading the prints to SD via the console port obviously improves the reliability but not the 

Having a quick look at the code here http://www.raspberrypi.org/forums/viewtopic.php?p=295695 it seems to be possible the BitBang an SD card via the GPIO pins on the Pi. The only thing we would need to do is use another pin or two (and maybe a fet) to assert the "card insert" pin on the MCU to "disconnect" and "reconnect" the SD card and have the MCU re-read the contents

We could even do this using CGODE assuming the user doesnt touch the LCD whilst we are uploading to the SDcard

I'm aware something like the Smoothieboard allows the MCU and Host device to access the SDcard at the "same time", but Smoothie firmware doesn't support auto bed leveling and a bunch of other cool features

I *think* I can design the hardware to make this work, but my software skills are lacking.. the last code I touched was VB6 and that wasnt very elegant

Or Option 2: Does anyone have any thoughts on emulating the SD card directly so we just connect the GPIO pins into the SDcard pins on the RAMPS based MCU?

**Comment by Atreidae** [comment]
Turns out the issue was I was pushing the RPi a little too hard. With my Webcam running at full 1920x1080 and timelapse hitting every 15 seconds. The serial buffer in the MCU didn't have time to fill before the RPi CPU became loaded down trying to deal with timelapse etc.
This causes the serial interface to slowdown which caused Marlin to throw an exception.

I've lessened the load on the RPi and I'm now printing happily.
